pkg/providers/cloudstack/validator.go

```diff
@@ -28,104 +30,153 @@ var restrictedUserCustomDetails = [...]string{
 		}
-		err := v.cmk.ValidateNetworkPresent(ctx, domainId, zone, zones, datacenterConfig.Spec.Account, len(zones) > 1)
+		endpoint := cmk.GetManagementApiEndpoint()
+		if endpoint != az.ManagementApiEndpoint {
+			return fmt.Errorf("cloudstack secret management url (%s) differs from cluster spec management url (%s)",
+				endpoint, az.ManagementApiEndpoint)
+		}
+
+		zoneId, err := cmk.ValidateZonePresent(ctx, az.CloudStackAvailabilityZone.Zone)
 		if err != nil {
-			return fmt.Errorf("checking network %v", err)
+			return err
+		}
+		az.CloudStackAvailabilityZone.Zone.Id = zoneId
+		if len(az.CloudStackAvailabilityZone.Zone.Network.Id) == 0 && len(az.CloudStackAvailabilityZone.Zone.Network.Name) == 0 {
+			return fmt.Errorf("zone network is not set or is empty")
+		}
+		if err := cmk.ValidateNetworkPresent(ctx, az.DomainId, az.CloudStackAvailabilityZone.Zone.Network, zoneId, az.Account, true); err != nil {
+			return err
 		}
 	}
 
 	logger.MarkPass("Datacenter validated")
 	return nil
 }
 
-func (v *Validator) validateDomainAndAccount(ctx context.Context, datacenterConfig *anywherev1.CloudStackDatacenterConfig) error {
-	if (datacenterConfig.Spec.Domain != "" && datacenterConfig.Spec.Account == "") ||
-		(datacenterConfig.Spec.Domain == "" && datacenterConfig.Spec.Account != "") {
-		return fmt.Errorf("both domain and account must be specified or none of them must be specified")
+func (v *Validator) generateLocalAvailabilityZones(ctx context.Context, datacenterConfig *anywherev1.CloudStackDatacenterConfig) error {
+	if datacenterConfig == nil {
+		return errors.New("CloudStack Datacenter Config is null")
 	}
 
-	if datacenterConfig.Spec.Domain != "" && datacenterConfig.Spec.Account != "" {
-		domain, errDomain := v.cmk.ValidateDomainPresent(ctx, datacenterConfig.Spec.Domain)
-		if errDomain != nil {
-			return fmt.Errorf("checking domain: %v", errDomain)
+	if len(datacenterConfig.Spec.Domain) > 0 {
+		cmk, ok := v.cmks[decoder.CloudStackGlobalAZ]
+		if !ok {
+			return fmt.Errorf("cannot find CloudStack profile for availability zone %s", decoder.CloudStackGlobalAZ)
 		}
-
-		errAccount := v.cmk.ValidateAccountPresent(ctx, datacenterConfig.Spec.Account, domain.Id)
-		if errAccount != nil {
-			return fmt.Errorf("checking account: %v", errAccount)
+		domain, err := cmk.ValidateDomainPresent(ctx, datacenterConfig.Spec.Domain)
+		if err != nil {
+			return err
+		}
+		if err := cmk.ValidateAccountPresent(ctx, datacenterConfig.Spec.Account, domain.Id); err != nil {
+			return err
+		}
+		for _, zone := range datacenterConfig.Spec.Zones {
+			availabilityZone := localAvailabilityZone{
+				CloudStackAvailabilityZone: &anywherev1.CloudStackAvailabilityZone{
+					CredentialsRef:        decoder.CloudStackGlobalAZ,
+					Domain:                datacenterConfig.Spec.Domain,
+					Account:               datacenterConfig.Spec.Account,
+					ManagementApiEndpoint: datacenterConfig.Spec.ManagementApiEndpoint,
+					Zone:                  zone,
```

Comment:
this seems to be missing a name, doesn't it? I think we need the name too for some of the error messages

Reply:
Adding Name as SetDefaults does